src/shared/Core/Trace2.cs

```diff
@@ -491,16 +645,26 @@ public override string ToJson()
 
     public override string ToNormalString()
     {
-        return BuildNormalString(Exe.ToLower());
+        return BuildNormalString(GetEventMessage());
+    }
+
+    public override string ToPerformanceString()
+    {
+        return BuildPerformanceString(GetEventMessage());
+    }
```

Comment:
I think we can simplify `Trace2Message` here by pulling `string GetEventMessage()` up to the abstract `Trace2Message` class, and then call `GetEventMessage()` inside of the `Build*String()` methods.

```patch
diff --git a/src/shared/Core.Tests/Trace2Tests.cs b/src/shared/Core.Tests/Trace2Tests.cs
index 9d6fed32..e6abc0ed 100644
--- a/src/shared/Core.Tests/Trace2Tests.cs
+++ b/src/shared/Core.Tests/Trace2Tests.cs
@@ -17,17 +17,12 @@ internal class TestTrace2Message : Trace2Message
 
     public override string ToJson()
     {
-        return "";
+        return string.Empty;
     }
 
-    public override string ToNormalString()
+    protected override string GetEventMessage()
     {
-        return "";
-    }
-
-    public override string ToPerformanceString()
-    {
-        return "";
+        return string.Empty;
     }
 }
 
diff --git a/src/shared/Core/Trace2.cs b/src/shared/Core/Trace2.cs
index 95d4a60b..ce2ba7f2 100644
--- a/src/shared/Core/Trace2.cs
+++ b/src/shared/Core/Trace2.cs
@@ -493,7 +493,6 @@ public abstract class Trace2Message
     public DateTimeOffset Time { get; set; }
 
     [JsonProperty("file", Order = 5)]
-
     public string File { get; set; }
 
     [JsonProperty("line", Order = 6)]
@@ -504,12 +503,10 @@ public abstract class Trace2Message
 
     public abstract string ToJson();
 
-    public abstract string ToNormalString();
-
-    public abstract string ToPerformanceString();
-
-    protected string BuildNormalString(string message)
+    public string ToNormalString()
     {
+        string message = GetEventMessage();
+
         // The normal format uses local time rather than UTC time.
         string time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
         string source = GetSource();
@@ -519,8 +516,10 @@ public abstract class Trace2Message
         return $"{time} {source,-33} {Event.ToString().ToSnakeCase()} {message}";
     }
 
-    protected string BuildPerformanceString(string message)
+    public string ToPerformanceString()
     {
+        string message = GetEventMessage();
+
         // The performance format uses local time rather than UTC time.
         var time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
         var source = GetSource();
@@ -544,6 +543,8 @@ public abstract class Trace2Message
         //     LF
         return $"{time} {source,-29}| d{Depth} | {Thread,-24} | {Event.ToString().ToSnakeCase(),-12} {sb} | {message}";
     }
+    
+    protected abstract string GetEventMessage();
 
     internal string BuildSpan(PerformanceFormatComponent component)
     {
@@ -643,17 +644,7 @@ public class VersionMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage()
     {
         return Exe.ToLower();
     }
@@ -677,17 +668,7 @@ public class StartMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage()
     {
         return string.Join(" ", Argv);
     }
@@ -711,17 +692,7 @@ public class ExitMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage()
     {
         return $"elapsed:{ElapsedTime} code:{Code}";
     }
@@ -751,21 +722,11 @@ public class ChildStartMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage(Id.ToString()));
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage($"ch{Id}"));
-    }
-
-    private string GetEventMessage(string id)
+    protected override string GetEventMessage()
     {
         var sb = new StringBuilder();
         AddDots(sb);
-        sb.Append($"[{id}] {string.Join(" ", Argv)}");
+        sb.Append($"[{Id}] {string.Join(" ", Argv)}");
         return sb.ToString();
     }
 }
@@ -794,21 +755,11 @@ public class ChildExitMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage(Id.ToString()));
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage($"ch{Id}"));
-    }
-
-    private string GetEventMessage(string id)
+    protected override string GetEventMessage()
     {
         var sb = new StringBuilder();
         AddDots(sb);
-        sb.Append($"[{id}] pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
+        sb.Append($"[{Id}] pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
         return sb.ToString();
     }
 }

```

Reply:
How in this case do we add the `ch` prefix to the process id in child start and exit event messages? For example, the last line of the patch will output `[1]` when the correct formatting for the performance format target is `ch[1]`.

Reply:
Can we have `GetEventMessage(..)` take a `Trace2FormatTarget`, and all other implementations ignore the arg, but this particular event adds the `ch` in front of the `Id` if `formatTarget == Trace2FormatTarget.Performance`?

```diff
-    private string GetEventMessage(string id)
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         var sb = new StringBuilder();
         AddDots(sb);
-        sb.Append($"[{id}] pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
+        if (formatTarget == Trace2FormatTarget.Performance)
+            sb.Append($"[ch{Id}]");
+        else
+            sb.Append($"[{Id}]");
+        sb.Append($" pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
         return sb.ToString();
     }
 }
```

Reply:
Updated patch:

```patch
diff --git a/src/shared/Core.Tests/Trace2Tests.cs b/src/shared/Core.Tests/Trace2Tests.cs
index 9d6fed32..7b3b15b0 100644
--- a/src/shared/Core.Tests/Trace2Tests.cs
+++ b/src/shared/Core.Tests/Trace2Tests.cs
@@ -17,17 +17,12 @@ internal class TestTrace2Message : Trace2Message
 
     public override string ToJson()
     {
-        return "";
+        return string.Empty;
     }
 
-    public override string ToNormalString()
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
-        return "";
-    }
-
-    public override string ToPerformanceString()
-    {
-        return "";
+        return string.Empty;
     }
 }
 
diff --git a/src/shared/Core/Trace2.cs b/src/shared/Core/Trace2.cs
index 95d4a60b..31367ee9 100644
--- a/src/shared/Core/Trace2.cs
+++ b/src/shared/Core/Trace2.cs
@@ -493,7 +493,6 @@ public abstract class Trace2Message
     public DateTimeOffset Time { get; set; }
 
     [JsonProperty("file", Order = 5)]
-
     public string File { get; set; }
 
     [JsonProperty("line", Order = 6)]
@@ -504,12 +503,10 @@ public abstract class Trace2Message
 
     public abstract string ToJson();
 
-    public abstract string ToNormalString();
-
-    public abstract string ToPerformanceString();
-
-    protected string BuildNormalString(string message)
+    public string ToNormalString()
     {
+        string message = GetEventMessage(Trace2FormatTarget.Normal);
+
         // The normal format uses local time rather than UTC time.
         string time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
         string source = GetSource();
@@ -519,8 +516,10 @@ public abstract class Trace2Message
         return $"{time} {source,-33} {Event.ToString().ToSnakeCase()} {message}";
     }
 
-    protected string BuildPerformanceString(string message)
+    public string ToPerformanceString()
     {
+        string message = GetEventMessage(Trace2FormatTarget.Performance);
+
         // The performance format uses local time rather than UTC time.
         var time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
         var source = GetSource();
@@ -545,6 +544,8 @@ public abstract class Trace2Message
         return $"{time} {source,-29}| d{Depth} | {Thread,-24} | {Event.ToString().ToSnakeCase(),-12} {sb} | {message}";
     }
 
+    protected abstract string GetEventMessage(Trace2FormatTarget formatTarget);
+
     internal string BuildSpan(PerformanceFormatComponent component)
     {
         var sb = new StringBuilder();
@@ -643,17 +644,7 @@ public class VersionMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         return Exe.ToLower();
     }
@@ -677,17 +668,7 @@ public class StartMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         return string.Join(" ", Argv);
     }
@@ -711,17 +692,7 @@ public class ExitMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage());
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage());
-    }
-
-    private string GetEventMessage()
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         return $"elapsed:{ElapsedTime} code:{Code}";
     }
@@ -751,21 +722,18 @@ public class ChildStartMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage(Id.ToString()));
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage($"ch{Id}"));
-    }
-
-    private string GetEventMessage(string id)
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         var sb = new StringBuilder();
         AddDots(sb);
-        sb.Append($"[{id}] {string.Join(" ", Argv)}");
+
+        if (formatTarget == Trace2FormatTarget.Performance)
+            sb.Append($"[ch{Id}]");
+        else
+            sb.Append($"[{Id}]");
+
+        sb.Append($" {string.Join(" ", Argv)}");
+        
         return sb.ToString();
     }
 }
@@ -794,21 +762,18 @@ public class ChildExitMessage : Trace2Message
             });
     }
 
-    public override string ToNormalString()
-    {
-        return BuildNormalString(GetEventMessage(Id.ToString()));
-    }
-
-    public override string ToPerformanceString()
-    {
-        return BuildPerformanceString(GetEventMessage($"ch{Id}"));
-    }
-
-    private string GetEventMessage(string id)
+    protected override string GetEventMessage(Trace2FormatTarget formatTarget)
     {
         var sb = new StringBuilder();
         AddDots(sb);
-        sb.Append($"[{id}] pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
+        
+        if (formatTarget == Trace2FormatTarget.Performance)
+            sb.Append($"[ch{Id}]");
+        else
+            sb.Append($"[{Id}]");
+        
+        sb.Append($" pid:{Pid} code:{Code} elapsed:{ElapsedTime}");
+        
         return sb.ToString();
     }
 }

```